### PR TITLE
Implement proper tracked motion model

### PR DESCRIPTION
**Pull request by peci1**
This PR fixes issue #365 . It adds proper support for tracked motion to Absolem and MARV robots (it could be extended to other tracked robots, too).

Depends on ignition-forks/dart#22, ignitionrobotics/ign-physics#267 and https://github.com/ignitionrobotics/ign-gazebo/pull/869.

### CODE REVIEW

**Comment by peci1** [comment]
Videos of the implementation (they only work in Chrome, I don't know why).

https://user-images.githubusercontent.com/182533/122688768-75ac2580-d21e-11eb-9947-67844b548381.mp4


https://user-images.githubusercontent.com/182533/122688763-71800800-d21e-11eb-80b0-0fc894cdd7de.mp4

https://user-images.githubusercontent.com/182533/122688779-8bb9e600-d21e-11eb-8e36-b23a035fb8b5.mp4


https://user-images.githubusercontent.com/182533/122688788-9c6a5c00-d21e-11eb-9e18-60d625bc8761.mp4

**Comment by peci1** [comment]
All the required upstream releases have been finished. I'll have a look at updating this PR next week.

Should I also give a try at updating other teams' tracked robots? (at least CSIRO DTR and HD2 robots off the top of my head).

**Comment by peci1** [comment]
Is `master` the correct target branch for this PR?

**Comment by azeey** [comment]
> Is `master` the correct target branch for this PR?

Yes, that's the correct branch.

**Comment by peci1** [comment]
I think this PR is ready to be reviewed.

Do not get confused by moving the `<enable_orientation>0</enable_orientation>` tag of IMU to the end of the `<imu>` tag - that's just to synchronize the SDFs to the output generated by `ign sdf`.

**Comment by peci1** [comment]
To test this PR:

```
ign launch src/subt/submitted_models/ctu_cras_norlab_absolem_sensor_config_1/launch/example.ign robotName:=X1 ros:=true

# in another console:

# raise the front flippers
rostopic pub -1 /X1/flippers_cmd_pos/front_right std_msgs/Float64 "data: -0.5" &
rostopic pub -1 /X1/flippers_cmd_pos/front_left std_msgs/Float64 "data: -0.5" &

# drive forward over the pallet
rostopic pub -1 /X1/cmd_vel geometry_msgs/Twist "linear:
  x: 0.5
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"

# or you can test using a gamepad (deadman + D-pad for flippers)

# see how smoothly the robot climbs up the pallet (no bumping or getting stuck caused by fake wheels)
```

**Review by azeey — COMMENTED** [review]
I tested both Absolum and MARV robots on the create. Absolum was able to climb over it without anye issues, but MARV got stuck with half of the robot on the crate and half on the ground. Is that expected?

I also noticed that the battery doesn't start draining when either of the robots are commanded to move. Currently only the `LinearBattery` system monitors the `JointVelocityCmd` and `JointForceCmd` components to detect when the battery should start draining. https://github.com/ignitionrobotics/ign-gazebo/blob/acac5b50f75f672d9a335471966cd3681ce7926c/src/systems/battery_plugin/LinearBatteryPlugin.cc#L399. Since the `TrackedVehicle` system doesn't make use of these components, `LinearBattery` is unable to start draining the battery. One thought is to create a component that gets attached to a model and triggers `LinearBattery` to drain. This would in addition to the current trigger components `LinearBattery` monitors. Then `TrackedVehicle` can attach this component on the model when it receives a motion command.

**Comment by peci1** [comment]
If MARV got stuck with its belly, it's expected. This happens also with the real robot. You need to help the robot with rear flippers to climb up the pallet.

File: submitted_models/ctu_cras_norlab_marv_sensor_config_1/model.sdf [hunk 1]
```diff
@@ -4806,200 +3206,7 @@
...
-        <wheel_radius>0.111</wheel_radius>
-      </wheel>
-      <wheel link_name="rear_right_flipper_wheel1">
-        <slip_compliance_lateral>0.0485</slip_compliance_lateral>
-        <slip_compliance_longitudinal>0</slip_compliance_longitudinal>
-        <wheel_normal_force>7.53525</wheel_normal_force>
-        <wheel_radius>0.1165</wheel_radius>
-      </wheel>
-    </plugin>
+    <plugin name="ignition::gazebo::systems::WheelSlip" filename="ignition-gazebo-wheel-slip-system"/>
```
[the remark this answers is not in the document]
Comment (peci1) [reply submitted_models/ctu_cras_norlab_marv_sensor_config_1/model.sdf]: Fixed.

**Comment by peci1** [comment]
Regarding the battery plugin and new gazebo component - what about adding component `BatteryDrainControl` which would be the alternative way of enabling / disabling battery discharge? This would also require changing the `LinearBatteryPlugin`.

Another alternative (or just a temporary solution until next gazebo release) is copy-pasting the `LinearBatteryPlugin` in subt repo and adjusting it to cooperate with the tracked vehicle plugin.

Yet another alternative - the battery plugin could get a list of topics to subscribe to and it could start draining as soon as any message is received on any of these topics. In this particular case, it should receive the list of `cmd_vel` topics of the track controllers (4 topics on MARV, 6 topics on Absolem).

Last, the battery plugin could check pose of the model and start draining when it moves more than a specified amount (this should account for the initial little falls models can do when they are inserted into the simulation).

**Comment by peci1** [comment]
Or maybe the best option would be adding start_draining/stop_draining topics or services to the battery plugin?

**Comment by peci1** [comment]
What about this hotfix for the battery issue, which could be upstreamed? https://github.com/peci1/subt/commit/1951c0a578c7099b9ac5d14e794af629adef5013

**Comment by azeey** [comment]
I like the idea of the battery plugin monitoring a list of topics. We'll need to make it message type agnostic though. I'm okay with the hotfix for expediency. @nkoenig can you confirm?

**Comment by peci1** [comment]
I've changed the battery draining topics to generic raw subscription. You can track and comment changes related to the battery hotfix in https://github.com/peci1/subt/pull/2 .

**Comment by peci1** [comment]
I've started implementing the better track model also for other robots: https://github.com/peci1/subt/pull/3

**Comment by nkoenig** [comment]
Here is a PR that adds topic support to the battery plugin in gazebo: https://github.com/ignitionrobotics/ign-gazebo/pull/1255.

**Comment by peci1** [comment]
I've updated this PR and peci1/subt#3 to integrate with ignitionrobotics/ign-gazebo#1255 . So the battery drain should be fixed now for all the tracked vehicles.

**Review by azeey — APPROVED** [review]
Works great! Thanks for your patience @peci1.

**Comment by peci1** [comment]
Thanks! I've submitted https://github.com/peci1/subt/pull/3 as #1048 .

**Comment by peci1** [comment]
@azeey Can I ask for upload of these models to Fuel?

**Comment by nkoenig** [comment]
Sure thing @peci1. I'll start looking into an update tomorrow.

**Comment by nkoenig** [comment]
The models on Fuel have been updated.

**Comment by peci1** [comment]
Thanks for the upload, @nkoenig , and for merging #1048, @azeey ! Would you please also find time for looking at #1047 and #1045?

**Pull request merged**